### DISCUSSION TITLE
Running a code when alarm rings

### DISCUSSION THREAD

**Discussion by Yasser85** [gdelataillade/alarm#263] [Q&A] [answered]
Hello
Thank you for sharing this awesome package.
I was making an app (Android app) where there is a need to run a code when the alarm rings without the needs of showing a full screen notification (because there is no need for that in my case, it did not work on all devices and also android is restricting it hardly).
Basically, I wanted to add a feature to the app that gives the user to make the device silent for a specific time then get out of silent mode  automatically (for better sleeping time or when in meeting ... etc.) in addition to regular alarms.
I thought of doing this in the package's Kotlin code but I don't have the required knowledge to do that.
I looked at `gdelataillade/alarm/alarm/AlarmService.kt` and thought it can be done by making another `AlarmService` class but with some modifying the code using `volumeService?.restorePreviousVolume(showSystemUI)` function, but I lost a bit.
I want to keep both functionalities: regular alarms and running a piece of code (generally or at least for periodic silent mode).
What can I do in this case?
Thanks in advance.

**Comment by gdelataillade** [comment]
Hi @Yasser85 

To implement the feature where the system volume is set to silent when the alarm is scheduled, you should modify the `AlarmPlugin` class, specifically in the `handleDelayedAlarm` method. You’ll need to use the `VolumeService.setVolume` method to adjust the volume settings.

When setting your alarm, if you want to avoid a full-screen notification, set the `androidFullScreenIntent` parameter to false. Additionally, if you want the alarm to produce sound, ensure the `volume` is set to a value greater than 0.

**Reply by Yasser85** [reply]
Thank you.
I will try to do it as soon as I can then I will let you know the result.

**Reply by Yasser85** [reply]
Hi
Sorry I couldn't work on it until now.
About `handleDelayedAlarm`, isn't it executed when the alarm being set? and not when the alarm rings?
If so then it will not achieve what I mentioned, which is making the phone un-silent when the alarm rings after the specific time.
Actually a few weeks before I posted the main question, I planned to make exactly what you said, even wrote a note about all the changes I need to `AlarmPlugin.kt`, but got to this exact point and stopped there. That is why I tried to look at `AlarmService.kt` and as I said "lost a bit" 😅.

Thank you.

**Reply by gdelataillade** [reply, marked as answer]
Ah, sorry I mixed that up in my head. You’re right—`handleDelayedAlarm` runs when the alarm is set, not when it goes off. If you want to run some code when the alarm starts ringing, that’s in `AlarmService`.

Just to make sure I got this straight: you want to drop the system volume to 0 when the alarm rings, and then bring it back up after a certain time, right?

**Reply by Yasser85** [reply]
No problem. 
Thank you for your answers.
Yes, this will do.
